Defaults and cache

wicheda opened the pull request:
Adding:

 - cache: all retrieved flags are cached by identity, and if a call fails, used instead. Can be turned off as well.
 - defaults: if a call fails and there is nothing in the cache, the defaults (set in code) will be returned instead.

Can be merged alongside https://github.com/Flagsmith/flagsmith/pull/2237

matthewelwell commented:
@wicheda do we know why the ubuntu-build workflow is failing here?

wicheda commented on FlagsmithClient/Classes/Flag.swift:
```diff
@@ -10,7 +10,7 @@ import Foundation
 A Flag represents a feature flag on the server.
 */
-public struct Flag: Decodable {
+public struct Flag: Codable {
```
It's Codable so we can store them and reload them (for the cache). Yes, Codable also includes Decodable so we don't lose anything.

wicheda commented on FlagsmithClient/Classes/Flagsmith.swift:
```diff
@@ -55,15 +73,34 @@ public class Flagsmith {
-        completion(result)
+        switch result {
+        case .success(let flags):
+          self.updateCache(flags: flags, forIdentity: identity)
```
True, it is always nil in the else block, but really we're just saying pass whatever identity what passed into this method, whether it is nil or not. I think that's safer since if for example someone refactored it and moved around the else block, they may miss the identity part and not pass it through.

matthewelwell replied:
I guess this is why it would be nice to have unit test coverage here, but that's another conversation! 

For readability's sake, I do think that I prefer to be explicit here, so I think we should either omit the `forIdentity` parameter (as I think it defaults to `nil`) or pass `nil` explicitly here. That being said, I'm not sure that my feelings on this one are strong enough to argue if you are keen to leave it as it is.

wicheda commented on FlagsmithClient/Classes/Flagsmith.swift:
```diff
@@ -103,8 +142,10 @@ public class Flagsmith {
-        let value = flags.first(where: {$0.feature.name == id})?.value
-        completion(.success(value?.stringValue))
+        var flag = flags.first(where: {$0.feature.name == id})
+        flag = self.getFlagUsingCacheAndDefaults(withID: id, flag: flag, forIdentity: identity)
```
Your interpretation is correct, I just thought it would be better for brevity's sake, but can change it.

wicheda commented on FlagsmithClient/Classes/Flagsmith.swift:
```diff
@@ -215,4 +257,62 @@ public class Flagsmith {
+  }
+
+  /// Return an array of flag for an identity, including the cached flags (if enabled) and the default flags when they are not already present in the passed array
+  private func getFlagsUsingCacheAndDefaults(flags:[Flag], forIdentity identity: String? = nil) -> [Flag] {
```
That's a business decision I guess, but happy to take a steer from you and @dabeeeenster . So the question is, when a user retrieves all flags, should we always add the cached and default flags to the returned list, or only do that in the case of a failed call?

wicheda commented:
Hi @dabeeeenster / @matthewelwell , I'm back from my trip and have found some time to finish this work. So, I have:

 - ensured that the cache / defaults are only ever used if a call fails, not merging it when we have a successful API call
 - added an optional TTL for the cache so that it is not used if it is too stale
 - added a skipAPI option, that will not call the API if there is a valid cache present

I've also made the stylistic changes you requested @matthewelwell . I could look at adding tests but noticed there aren't currently any tests at the Flagsmith level yet, so it may be a fairly lengthy task to create those since it will be quite tricky to start making real calls, and simulating failed connections etc. I'd have to schedule that over the next couple of weeks if you did want it, so let me know.

wicheda commented on Example/FlagsmithClient/AppDelegate.swift:
```diff
@@ -18,7 +18,21 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+                                     Flag(featureName: "font_size", intValue:12, enabled: true),
+                                     Flag(featureName: "my_name", stringValue:"Testing", enabled: true)]
     
+    // set cache on / off (defaults to on)
```
Yes, fair point.

wicheda commented on FlagsmithClient/Classes/Flagsmith.swift:
```diff
@@ -42,8 +42,35 @@ public class Flagsmith {
+  private let NIL_IDENTITY_KEY = "nil-identity"
+  
+  /// Cached flags to fall back on if an API call fails, by identity
+  private var cachedFlags: [String:[Flag]] = [:]
```
@matthewelwell , if we want to integrate the cacheing at a deeper level, I'll have to look into that next week and see what is feasible with the existing HTTP architecture, but that may be doable.

matthewelwell replied:
Yes, that would be great. Let me know if you want to chat. You can catch me on Slack as needed.

matthewelwell commented:
superseded by #34
